### PR TITLE
Improve completions to suggest available grammars and elements within those grammars

### PR DESCRIPTION
**Pull request by rafaelbey**
As part of completions, on empty lines, this will suggest the different grammars available (###Grammar), and also elements that grammar support, like Database within the Relational grammar.

### CODE REVIEW

File: legend-engine-ide-lsp-default-extensions/pom.xml [hunk 1]
```diff
@@ -51,6 +51,11 @@
                 <scope>import</scope>
                 <version>${legend.engine.version}</version>
             </dependency>
+            <dependency>
+                <groupId>junit</groupId>
+                <artifactId>junit</artifactId>
+                <version>4.13</version>
+            </dependency>
```
Comment (rafaelbey) [inline legend-engine-ide-lsp-default-extensions/pom.xml:58]: Intellij complains about the version engine imports (4.13.1), and I found is a bug on those patch versions.  SO suggested to just use 4.13.

File: legend-engine-ide-lsp-server/src/test/java/org/finos/legend/engine/ide/lsp/server/TestLegendTextDocumentService.java [hunk 2]
```diff
@@ -86,6 +91,56 @@ public void testNoKeywordHighlighting() throws Exception
...
+        // with ### preceding
+        asserter.accept(new Position(9, 3), Set.of("TestGrammar", "EmptyGrammar"));
+        // with ###T preceding
+        asserter.accept(new Position(9, 4), Set.of("TestGrammar", "EmptyGrammar"));
+    }
+
+    private static void assertExpectedCompletionItems(String uri, TextDocumentService service, Position position, Set<String> expected) throws InterruptedException, ExecutionException
+    {
+
+    }
```
[the remark this answers is not in the document]
Comment (rafaelbey) [reply legend-engine-ide-lsp-server/src/test/java/org/finos/legend/engine/ide/lsp/server/TestLegendTextDocumentService.java:142]: Dont need.  Will remove on next commit.  I was exploring options to write the asserts.

File: legend-engine-ide-lsp-text-tools/src/main/java/org/finos/legend/engine/ide/lsp/text/LineIndexedText.java [hunk 3]
```diff
@@ -269,9 +269,9 @@ public boolean isValidInterval(int line, int startColumn, int endColumn)
      */
     public boolean isValidInterval(int startLine, int startColumn, int endLine, int endColumn)
     {
-        return (startLine == endLine) ?
-                isValidInterval(startLine, startColumn, endColumn) :
-                (startLine < endLine) && isValidColumn(startLine, startColumn) && isValidColumn(endLine, endColumn);
+        return (startLine == endLine && isValidInterval(startLine, startColumn, endColumn))
+                ||
+                ((startLine < endLine) && isValidColumn(startLine, startColumn) && isValidColumn(endLine, endColumn));
```
[the remark this answers is not in the document]
Comment (rafaelbey) [reply legend-engine-ide-lsp-text-tools/src/main/java/org/finos/legend/engine/ide/lsp/text/LineIndexedText.java]: Im reverting this.  I was experimenting with few ideas to try to make the code work with intervals that are not inclusive, but forgot to fully clean up the code after discarding these.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/PureLSPGrammarExtension.java [hunk 4]
```diff
@@ -61,32 +72,26 @@
...
 public class PureLSPGrammarExtension extends AbstractLegacyParserLSPGrammarExtension
 {
     private static final Logger LOGGER = LoggerFactory.getLogger(PureLSPGrammarExtension.class);
 
+    private static final Set<String> TYPES = Set.of(
+            "Association",
+            "Class",
+            "Enum",
+            "function",
+            "import",
```
[the remark this answers is not in the document]
Comment (rafaelbey) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/PureLSPGrammarExtension.java:87]: I did not know what name to give to this variable, but this is to drive completions.  Native function is something users never will be able to define on the legend grammar, while imports is something they use often.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/ServiceLSPGrammarExtension.java [hunk 5]
```diff
@@ -102,7 +102,7 @@ public class ServiceLSPGrammarExtension extends AbstractSectionParserLSPGrammarE
                     "  }\n" +
                     "}\n");
 
-    private JsonMapper resultMapper;
+    private volatile JsonMapper resultMapper;
```
[the remark this answers is not in the document]
Comment (rafaelbey) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/ServiceLSPGrammarExtension.java]: This leaked from my next change on concurrency, when trying to reduce sync blocks.  Reverting from this PR.

File: legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/PureLSPGrammarExtension.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevin-m-knight-gs) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/PureLSPGrammarExtension.java:87]: We should definitely change the variable name, then. If these are primarily to drive completions, then maybe something like COMPLETABLES.

[the remark this answers is not in the document]
Comment (rafaelbey) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/PureLSPGrammarExtension.java:87]: I rename it to SUGGESTABLE_KEYWORDS.  Let me know what you think.

[the remark this answers is not in the document]
Comment (kevin-m-knight-gs) [reply legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/PureLSPGrammarExtension.java:87]: I think that's a better name than my suggestion.

**Review by kevin-m-knight-gs — APPROVED** [review]

**Pull request merged**